walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,140 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
+  /**
+   * Returns all {@code TimeRange}s that satisfies the request constraints.
+   *
+   * Has O(n*m) time complexity, where n is the number of attendees in the request and m is the
+   * size of the {@code events} parameter.
+   *
+   * @param events Collection of already scheduled {@code Event}s for the day.
+   * @param request {@code MeetingRequest} containing all restraints for this query.
+   * @return A Collection containing all {@code TimeRange}s that satisfies the constraints
+   * specified by {@code request}.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<ArrayList<TimeRange>> attendeeAvailabilities =
+        getAllAttendeeAvailabilities(events, request.getAttendees());
+    ArrayList<TimeRange> availableTimes = allTimeRangesIntersection(attendeeAvailabilities);
+    availableTimes = removeTimesBelowDuration(availableTimes, request.getDuration());
+    return availableTimes;
+  }
+
+  /**
+   * Gets a list of all the available times that each attendee in attendees is available in the day.
+   *
+   * @param events {@code Event} Collection for the day.
+   * @param attendees The collection of attendees which we wish to find all the availabilities of.
+   * @return An ArrayList where each element is the list of times that a unique attendee is
+   *         available.
+   */
+  private ArrayList<ArrayList<TimeRange>> getAllAttendeeAvailabilities(
+      Collection<Event> events, Collection<String> attendees) {
+    ArrayList<ArrayList<TimeRange>> attendeeAvailabilities = new ArrayList<ArrayList<TimeRange>>();
+    for (String attendee : attendees) {
+      attendeeAvailabilities.add(getAttendeeAvailability(events, attendee));
+    }
+    return attendeeAvailabilities;
+  }
+
+  /**
+   * Gets the available times in a day that the given attendee is available.
+   *
+   * @param events {@code Event} Collection for the day.
+   * @param attendee Person who we wish to find the available times.
+   * @return {@code TimeRange} ArrayList where the attendee is available.
+   */
+  private ArrayList<TimeRange> getAttendeeAvailability(Collection<Event> events, String attendee) {
+    ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>();
+
+    // Add TimeRanges over the entire day, skipping over the events that contains the attendee.
```

Comment:
Events that contain the attendee or events that  _don't_ contain the attendee?

Reply:
Events that contain the attendee. I'll try to make the comment clearer. 👍